ably/realtime_channel_integration_test.go

```diff
@@ -294,6 +294,37 @@ func TestRealtimeChannel_ShouldSetProvidedReadLimit(t *testing.T) {
 	assert.Equal(t, int64(2048), client.Connection.ReadLimit())
 }
 
+func TestRealtimeChannel_SetsNoLimitIfServerNoLimits(t *testing.T) {
+	in := make(chan *ably.ProtocolMessage, 1)
+	out := make(chan *ably.ProtocolMessage, 16)
+
+	_, c := ablytest.NewRealtime(ably.WithDial(MessagePipe(in, out)))
+
+	connDetails := ably.ConnectionDetails{
+		ClientID:           "id1",
+		ConnectionKey:      "foo",
+		MaxFrameSize:       12,
+		MaxInboundRate:     14,
+		MaxMessageSize:     0,
+		ConnectionStateTTL: ably.DurationFromMsecs(time.Minute * 2),
+		MaxIdleInterval:    ably.DurationFromMsecs(time.Second),
+	}
+
+	in <- &ably.ProtocolMessage{
+		Action:            ably.ActionConnected,
+		ConnectionID:      "connection-id-1",
+		ConnectionDetails: &connDetails,
+	}
+
+	// Wait for a little bit for things to settle
+	time.Sleep(1 * time.Second)
+
+	// Check that the connection read limit is the default - due to websocket.go
+	// Only allowing the value to be set if the connection is a certain type
+	// We'll just check -1 here
+	assert.Equal(t, int64(-1), c.Connection.ReadLimit())
```

Comment:
`c.Connection.ReadLimit()` should return `64kb`. It only sets `-1` if there is error while setting the limit. When it's set to `-1`, limit is `32kb` which is default `nhooyr/websocket` readlimit. Currently, there is an error while setting the limit, because it doesn't use `websocketConn` as such -> https://github.com/ably/ably-go/blob/cc9e260bf265f89cb42fd8f19b62b904fef2d1c8/ably/websocket.go#L122
Seems we need to test it as a integration test without mocking `websocket_conn`

Reply:
@AndyTWF is it fine if I can write integration test for this? seems mocking won't work in current case : (

Reply:
@sacOO7 I'm not sure how we achieve this without mocking - unless there's a way to tell sandbox to send 0, which I doubt we can to just for this test?

Reply:
I'm thinking ...

Reply:
One of the option I can think of is, updating `setConnectionReadLimit` method
```
func setConnectionReadLimit(c conn, readLimit int64) error {
	verboseConn, ok := c.(verboseConn)
	if !ok {
		return errors.New("cannot set readlimit for connection, connection does not use verboseConn")
	}
	websocketConn, ok := verboseConn.conn.(*websocketConn)
	if !ok {
		return errors.New("cannot set readlimit for connection, connection does not use nhooyr.io/websocket")
	}
	websocketConn.conn.SetReadLimit(readLimit)
	return nil
}
```
in a way that it recursively finds `websocketConn` as a part of supplied param `conn`. Then we set the readlimit on it. This doesn't put restriction of being `verboseConn` before `websocketConn`.


Then, we can have integration test with `DialFunc` that wraps `DialWebsocket conn`
```
func TestRealtimeConn_SendErrorReconnects(t *testing.T) {
	sendErr := make(chan error, 1)
	closed := make(chan struct{}, 1)
	allowDial := make(chan struct{})

	dial := DialFunc(func(p string, url *url.URL, timeout time.Duration) (ably.Conn, error) {
		<-allowDial
		ws, err := ably.DialWebsocket(p, url, timeout)
		if err != nil {
			return nil, err
		}
		return connMock{
			SendFunc: func(m *ably.ProtocolMessage) error {
				select {
				case err := <-sendErr:
					return err
				default:
					return ws.Send(m)
				}
			},
			ReceiveFunc: ws.Receive,
			CloseFunc: func() error {
				closed <- struct{}{}
				return ws.Close()
			},
		}, nil
	})

	app, c := ablytest.NewRealtime(ably.WithDial(dial))
	defer safeclose(t, ablytest.FullRealtimeCloser(c), app)

	allowDial <- struct{}{}

	err := ablytest.Wait(ablytest.ConnWaiter(c, c.Connect, ably.ConnectionEventConnected), nil)
	assert.NoError(t, err)

	// Cause a send error; expect message to be enqueued and transport to be
	// closed.

	sendErr <- errors.New("fail")

	publishErr := make(chan error, 1)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	go func() {
		e := c.Channels.Get("test").Publish(ctx, "test", nil)
		publishErr <- e
	}()

	ablytest.Instantly.Recv(t, nil, closed, t.Fatalf)
	ablytest.Instantly.NoRecv(t, nil, publishErr, t.Fatalf)

	// Reconnect should happen instantly as a result of transport closure.
	allowDial <- struct{}{}

	// After reconnection, message should be published.
	ablytest.Soon.Recv(t, &err, publishErr, t.Fatalf)
	assert.NoError(t, err)
}
```

Reply:
or if we don't want to modify `setConnectionReadLimit` method, then we need to wrap result from 
```
 ws, err := ably.DialWebsocket(p, url, timeout)
```
inside
 `verboseConn`

Reply:
@AndyTWF you can decide if all of those efforts are worth it in the end haha : )

Reply:
Okay, it seems if we can modify `setConnectionReadLimit`, we have option to use 
https://github.com/ably/ably-go/blob/cc9e260bf265f89cb42fd8f19b62b904fef2d1c8/ably/ably_test.go#L461

Reply:
The last option probably sounds like the path of least resistance?

Reply:
Yeah .. seems like it

Reply:
Okay, I updated the code. Also, unwrapping approach feels a bit cleaner :  )